DBZ-6652 Blog post about JMX signaling and notification (#957)
debezium/debezium.github.io: _posts/2023-09-25-Debezium-JMX-signaling-and-notifications.adoc

----

or I would suggest to use **https://github.com/kcctl/kcctl[kcctl]** tool to interact with Kafka Connect. It is a modern and intuitive command line client for Kafka Connect.

First of all we need to create a configuration context to connect with Kafka Connect

[source, bash]
----
kcctl config set-context local --cluster http://localhost:8083
----

then we can register the connector running the following command

[source, bash]
----
kcctl apply -f postgres-jmx.json
----

we can now get the logs of connect container

[source, bash]
----

> docker logs f1d49fb79dba <1>

Reviewer: you can use `docker compose logs connect` instead

----
<1> This is the id of the connect container that we got before, running the **docker** ps command

and check that the connector is started streaming events

[source, text]
----
INFO   Postgres|dbserver1|streaming  Starting streaming   [io.debezium.pipeline.ChangeEventSourceCoordinator]
----

=== Prepare database for incremental snapshot
Since incremental snapshot still requires to have the `signal.data.collection` we need to create the signal table on our postgres database.

[NOTE]
The data collection is not required if you are using MySql with GTIds and `read.only` set to true.

To create the signal table we need to connect to our postgres instance. We can use the `psql` client inside the postgres container.

[source, bash]
----

> docker exec -it e61116f22f9d bash <1>

Reviewer: and same here: ` docker compose exec -it postgres bash` (or directly `docker compose exec -it postgres psql -h localhost -d postgres -U postgres`)

---
layout: post
title:  "Debezium signaling and notifications - Part 3: JMX channel"
date:   2023-09-25 11:11:11 +0100
tags: [ debezium, features, notifications, signaling, integration ]
featured: true
author: mfvitale
---
Welcome to the third installment of our series on Debezium Signaling and Notifications.
In this article, we continue our exploration of Debezium signaling and notifications. In particular, we will delve into how to enable and manage these features using the JMX channel.

> We will also explore how to send signal and get notification through REST api leveraging on Jolokia.

Reviewer 2: ```suggestion
We will also explore how to send signals and get notifications through the REST API leveraging Jolokia.
```

+++<!-- more -->+++

== Interact with Debezium via JMX

> JMX stands for Java Management Extensions, it is a Java technology used for managing and monitoring Java applications. It provides a standardized way to monitor application performance, configure settings, and interact with running Java applications using various management tools and clients. JMX is particularly useful for managing and monitoring complex, distributed, and enterprise-level Java applications.

Reviewer 2: ```suggestion
JMX stands for Java Management Extensions, a Java technology for managing and monitoring Java applications. It provides a standardized way to monitor application performance, configure settings, and interact with running Java applications using various management tools and clients. JMX is particularly useful for managing and monitoring complex, distributed, and enterprise-level Java applications.
```

=== Enable signaling  via JMX channel

> Signaling in Debezium is all about triggering actions to execute operation during normal execution. As discussed in previous articles, Debezium offers different out-of-the-box signal channels.
> In this article we will focus on JMX channel.

Reviewer 2: ```suggestion
Signaling in Debezium is about triggering actions to execute operations during normal execution. As discussed in previous articles, Debezium offers different out-of-the-box signal channels.
In this article, we will focus on the JMX channel.
```

> To start using JMX signal channel we need to:

Reviewer 2: ```suggestion
To start using the JMX signal channel, we need to:
```

* Enable JMX server on Kafka Connect service
* Add `jmx` to `signal.enabled.channels` connector configuration property
* Use a JMX client to connect to JMX server to send signal.

> Debezium expose the signal MBean named `debezium.<connector-type>:type=management,context=signals,server=<server>`, it exposes `signal` operations that accept three parameters:

Reviewer 2: ```suggestion
Debezium exposes the signal MBean named `debezium.<connector-type>:type=management,context=signals,server=<server>`.
This bean exposes `signal` operations that accept three parameters:
```

* p0: The id of the signal.

* p1: The type of the signal, for example, execute-snapshot.

* p2: A JSON data field that contains additional information about the specified signal type.

=== Enable notifications via JMX channel

> Notifications play a pivotal role in keeping you informed about what happens in Debezium. Accessing notification via JMX channel permits you to easily monitor Debezium, for example the progress of an incremental snapshot.

Reviewer 2: ```suggestion
Notifications are pivotal in informing you about what happens in Debezium. Accessing notifications via the JMX channel permits you to monitor Debezium easily, for example, the progress of an incremental snapshot.
```

> To start using JMX notification channel we need to:

Reviewer 2: ```suggestion
To start using the JMX notification channel, we need to:
```

* Enable JMX server on Kafka Connect service
* Add `jmx` to `notification.enabled.channels` connector configuration property
* Use a JMX client to connect to JMX server to access to notifications.

> Debezium expose the notification MBean named `debezium.<connector-type>:type=management,context=notifications,server=<server>`, it provides `Notification` bean that is a list of JMX `CompositeData` type with following attributes:

Reviewer 2: ```suggestion
Debezium exposes the notification MBean named `debezium.<connector-type>:type=management,context=notifications,server=<server>`.
This bean provides a `Notification` bean that contains a list of JMX `CompositeData` types with the following attributes:
```

|===
|Property |Description

|id
|A unique identifier that is assigned to the notification. For incremental snapshot notifications, the `id` is the same sent with the `execute-snapshot` signal.

|aggregate_type
|The data type of the aggregate root to which a notification is related.
In domain-driven design, exported events should always refer to an aggregate.

|type
|Provides status information about the event specified in the `aggregate_type` field.

|additional_data
|A Map<String,String> with detailed information about the notification.
|===

> But let's put our hands on and let's se how to send an incremental snapshot and receive notification about its progress through JMX channel.

Reviewer 2: ```suggestion
Let's take a moment and see how to send an incremental snapshot and to receive notifications about its progress through the JMX channel.
```

=== Send incremental snapshot signal through JMX channel

> For this example, we will use debezium docker images with PostgreSQL database.

Reviewer 2: ```suggestion
We will use the Debezium docker images with a PostgreSQL database for this example.
```

ccb502882928   quay.io/debezium/zookeeper:2.4   "/docker-entrypoint.…"   4 seconds ago   Up 2 seconds   0.0.0.0:2181->2181/tcp, 0.0.0.0:2888->2888/tcp, 0.0.0.0:3888->3888/tcp             deploy-zookeeper-1
----

At this point all services are up and running, so we can register the connector through the following configuration

[source, json]
----
{
  "name": "inventory-connector",
  "config": {
    "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
    "tasks.max": "1",
    "database.hostname": "postgres",
    "database.port": "5432",
    "database.user": "postgres",
    "database.password": "postgres",
    "database.server.id": "184054",
    "database.dbname": "postgres",
    "topic.prefix": "dbserver1",
    "snapshot.mode": "NEVER",
    "schema.history.internal.kafka.bootstrap.servers": "kafka:9092",
    "schema.history.internal.kafka.topic": "schema-changes.inventory",
    "signal.enabled.channels": "source,jmx", <1>
    "signal.data.collection": "inventory.debezium_signal", <2>
    "notification.enabled.channels": "jmx"
  }
}
----

> <1> This configuration enables **source** and **jmx** channel. Even if we will use JMX to send the signal, since we want to execute an incremental snapshot, the **source** signal is still required because Debezium needs to use signal table to watermark the db log for events decoupling.

Reviewer 2: ```suggestion
<1> This configuration enables the **source** and **jmx** channels. Even if we only want to use JMX to send signals to execute an incremental snapshot, the **source** signal is still required because Debezium needs to use signal table to watermark the db log for event deduplication.
```

----

or I would suggest to use **https://github.com/kcctl/kcctl[kcctl]** tool to interact with Kafka Connect. It is a modern and intuitive command line client for Kafka Connect.

First of all we need to create a configuration context to connect with Kafka Connect

[source, bash]
----
kcctl config set-context local --cluster http://localhost:8083
----

then we can register the connector running the following command

[source, bash]
----
kcctl apply -f postgres-jmx.json
----

we can now get the logs of connect container

[source, bash]
----
docker logs connect
----

and check that the connector is started streaming events

[source, text]
----
INFO   Postgres|dbserver1|streaming  Starting streaming   [io.debezium.pipeline.ChangeEventSourceCoordinator]
----

=== Prepare database for incremental snapshot

> Since incremental snapshot still requires to have the `signal.data.collection` we need to create the signal table on our postgres database.

Reviewer 2: ```suggestion
Since incremental snapshots require the `signal.data.collection` to be defined, we need to create the signal table on our postgres database.
```

[NOTE]

> The data collection is not required if you are using MySql with GTIds and `read.only` set to true.

Reviewer 2: ```suggestion
The data collection is not required when using MySql with GTIDs and  `read.only` set to true.
```

[source, bash]
----
docker exec -it postgres bash
----

Once inside the container we can connect to postgres instance with

[source, bash]
----
psql -h localhost -d postgres -U postgres
----

[NOTE]
password is *postgres*

then we can check that there are already some tables inside the *inventory* schema

[source, bash]
----
\dt inventory.*
----

the command should return something similar to

[source, bash]
----
                List of relations
  Schema   |       Name       | Type  |  Owner
-----------+------------------+-------+----------
 inventory | customers        | table | postgres
 inventory | geom             | table | postgres
 inventory | orders           | table | postgres
 inventory | products         | table | postgres
 inventory | products_on_hand | table | postgres
 inventory | spatial_ref_sys  | table | postgres
(6 rows)

----

We need to create the signal table with the following command:

[source, bash]
----
CREATE TABLE inventory.debezium_signal (id VARCHAR(42) PRIMARY KEY, type VARCHAR(32) NOT NULL, data VARCHAR(2048) NULL);
----

=== Send incremental snapshot signal

> To send the signal through JMX channel we need to connect to JMX server. We use the https://github.com/jiaqi/jmxterm[jmxterm] client so, after downloading it, we can run it

Reviewer 2: ```suggestion
We must connect to the JMX server to send the signal through the JMX channel. We use the https://github.com/jiaqi/jmxterm[jmxterm] client, so after downloading it, we can run it
```

[source,bash]
----
java -jar jmxterm-1.0.4-uber.jar <1>

open localhost:9012 <2>

beans -d debezium.postgres <3>

run -b debezium.postgres:context=signals,server=dbserver1,type=management signal 12345 execute-snapshot {"data-collections":["inventory.orders"],"type":"INCREMENTAL"} <4>
----
<1> Run the jmxterm client
<2> Open a connection to JMX Server
<3> Search for bean under **debezium.postgres** domain
<4> Execute the **signal** operation to execute an incremental snapshot of the **inventory.orders** table.

=== Check data

> After that we want to check that all data from *orders* table has been correctly captured in its corresponding Kafka topic.

Reviewer 2: ```suggestion
After that, we want to check that all the data from the *orders* table has been correctly captured in its corresponding Kafka topic.
```

            "type": "string",
            "optional": false,
            "field": "id"
          },
          {
            "type": "int64",
            "optional": false,
            "field": "total_order"
          },
          {
            "type": "int64",
            "optional": false,
            "field": "data_collection_order"
          }
        ],
        "optional": true,
        "name": "event.block",
        "version": 1,
        "field": "transaction"
      }
    ],
    "optional": false,
    "name": "dbserver1.inventory.orders.Envelope",
    "version": 1
  },
  "payload": {
    "before": null,
    "after": {
      "id": 10004,
      "order_date": 16852,
      "purchaser": 1003,
      "quantity": 1,
      "product_id": 107
    },
    "source": {
      "version": "2.4.0-SNAPSHOT",
      "connector": "postgresql",
      "name": "dbserver1",
      "ts_ms": 1695631605204,
      "snapshot": "incremental",
      "db": "postgres",
      "sequence": "[\"34837776\",\"34837776\"]",
      "schema": "inventory",
      "table": "orders",
      "txId": null,
      "lsn": null,
      "xmin": null
    },
    "op": "r",
    "ts_ms": 1695631605204,
    "transaction": null
  }
}
----

> That's all! We have sent an incremental snapshot signal using JMX channel.

Reviewer 2: ```suggestion
That's all! We have sent an incremental snapshot signal using the JMX channel.
```

=== Monitor incremental snapshot progress through JMX channel

> Since we have executed an incremental snapshot we can now read the notifications produced by Debezium via JMX channel.

Reviewer 2: ```suggestion
Since we have executed an incremental snapshot, we can now read the notifications produced by Debezium via the JMX channel.
```

We used the following configuration to register our connector

[source, json]
----
{
  "name": "inventory-connector",
  "config": {
    "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
    "tasks.max": "1",
    "database.hostname": "postgres",
    "database.port": "5432",
    "database.user": "postgres",
    "database.password": "postgres",
    "database.server.id": "184054",
    "database.dbname": "postgres",
    "topic.prefix": "dbserver1",
    "snapshot.mode": "NEVER",
    "schema.history.internal.kafka.bootstrap.servers": "kafka:9092",
    "schema.history.internal.kafka.topic": "schema-changes.inventory",
    "signal.enabled.channels": "source,jmx",
    "signal.data.collection": "inventory.debezium_signal",
    "notification.enabled.channels": "jmx" <1>
  }
}
----
<1> This configuration enables **jmx** notification channel.

> To access the notifications we need to connect again to JMX server. So as we did for the signal we will use `jmxterm`

Reviewer 2: ```suggestion
To access the notifications, we need to connect again to the JMX server. So as we did for the signal, we will use `jmxterm`
```

      key = connector_name;
      value = dbserver1;
     };
    ( maximum_key ) = {
      key = maximum_key;
      value = 10004;
     };
    ( last_processed_key ) = {
      key = last_processed_key;
      value = 10004;
     };
    ( data_collections ) = {
      key = data_collections;
      value = inventory.orders;
     };
   };
  aggregateType = Incremental Snapshot;
  id = 12345;
  type = IN_PROGRESS;
 }, <3>
{
  additionalData = {
    ( scanned_collection ) = {
      key = scanned_collection;
      value = inventory.orders;
     };
    ( connector_name ) = {
      key = connector_name;
      value = dbserver1;
     };
    ( total_rows_scanned ) = {
      key = total_rows_scanned;
      value = 4;
     };
    ( status ) = {
      key = status;
      value = SUCCEEDED;
     };
    ( data_collections ) = {
      key = data_collections;
      value = inventory.orders;
     };
   };
  aggregateType = Incremental Snapshot;
  id = 12345;
  type = TABLE_SCAN_COMPLETED;
 }, <4>
{
  additionalData = {
    ( connector_name ) = {
      key = connector_name;
      value = dbserver1;
     };
   };
  aggregateType = Incremental Snapshot;
  id = 12345;
  type = COMPLETED;
 } <5>
];
----

> <1> This is a notification from initial snapshot that as the status `SKIPPED` since our connector is configured with `"snapshot.mode": "NEVER"`

Reviewer 2: ```suggestion
<1> This is a notification from the initial snapshot with the status `SKIPPED` since our connector is configured with `"snapshot.mode": "NEVER"`
```

<2> This is the notification about the incremental snapshot start

> <3> This notification tells that the snapshot of table `inventory.orders` is in progress and provide some useful information about the last processed key and the max key. In this particular example we have only one `in progress` notification but depending on your table size and `snapshot.fetch.size` you can get more.

Reviewer 2: ```suggestion
<3> This notification tells that the table `inventory.orders` snapshot is in progress and provides useful information about the last processed and max keys. In this example, we have only one `in progress` notification, but depending on your table size and `snapshot.fetch.size`, you can get more.
```

> <4> This notification tells that the snapshot for a specific table has completed and provides also information about the total row processed.

Reviewer 2: ```suggestion
<4> This notification tells that the snapshot for a specific table has been completed and provides information about the total rows processed.
```

<5> This is the last notification that we have for this example and tells that the entire incremental snapshot progress has completed.

[NOTE]

> JMX offers also the possibility to produce its own notification, Debezium will also produce these notifications.
> You can, depending on your jmx client, subscribe to these notification, so you can immediately receive it without the need to poll the *Notification* bean.

Reviewer 2: ```suggestion
JMX also offers the possibility to produce its own notifications. Debezium will also produce these notifications.
Depending on your JMX client, you can subscribe to these notifications, so you can immediately receive them without polling the *Notification* bean.
```

== Leveraging Jolokia for REST-based Signaling and Notifications

> Jolokia is a powerful tool that allows you to interact with a JMX server and expose it via REST.
> Using it we can interact with Debezium via REST leveraging the signaling and notifications JMX beans.
> In this way, you can send signals and receive notifications seamlessly and with more familiar REST api.

Reviewer 2: ```suggestion
Jolokia is a powerful tool that allows you to interact with a JMX server and expose it via REST.
Using it, we can interact with Debezium via REST, leveraging the signaling and notifications JMX beans.
In this way, you can send signals and receive notifications seamlessly and with a more familiar REST api.
```

      "type": "STARTED",
      "aggregateType": "Incremental Snapshot"
    },
    {
      "additionalData": {
        "last_processed_key": "109",
        "current_collection_in_progress": "inventory.products",
        "connector_name": "dbserver1",
        "maximum_key": "109",
        "data_collections": "inventory.products"
      },
      "id": "12345",
      "type": "IN_PROGRESS",
      "aggregateType": "Incremental Snapshot"
    },
    {
      "additionalData": {
        "scanned_collection": "inventory.products",
        "connector_name": "dbserver1",
        "total_rows_scanned": "9",
        "status": "SUCCEEDED",
        "data_collections": "inventory.products"
      },
      "id": "12345",
      "type": "TABLE_SCAN_COMPLETED",
      "aggregateType": "Incremental Snapshot"
    },
    {
      "additionalData": {
        "connector_name": "dbserver1"
      },
      "id": "12345",
      "type": "COMPLETED",
      "aggregateType": "Incremental Snapshot"
    }
  ],
  "timestamp": 1695652278,
  "status": 200
}

----

You can see that now we have also the notification about `inventory.products` table incremental snapshot that we have sent through REST api

== Conclusion

> In this third installment of our series on Debezium Signaling and Notifications, we've learned how to enable and manage both signaling and notifications using JMX and Jolokia.
> Signaling empowers you to dynamically control Debezium's behavior, while notifications keep you informed about critical events. By harnessing these capabilities along with Jolokia,
> you can effectively manage, monitor, and interact with your data streaming workflows, ensuring that you stay in control of Debezium at all times.

Reviewer 2: ```suggestion
In this third installment of our series on Debezium Signaling and Notifications, we've learned how to enable and manage both signaling and notifications using JMX and Jolokia.
Signaling lets you dynamically control Debezium's behavior, while notifications keep you informed about critical events. By harnessing these capabilities along with Jolokia,
you can effectively manage, monitor, and interact with your data streaming workflows, ensuring that you always control Debezium.
```
